File: 2023/day_8/part_1.rb
require './input_reader'
require 'byebug'

instructions = nil
network = {}
index = 0
MAP_REGEX = /\A(?<node>.+) = \((?<left>.+), (?<right>.+)\)\z/

InputReader.each_line(File.join(File.dirname(__FILE__), "#{ARGV[0] || 'input'}.txt")) do |line|
  case index
  when 0
    instructions = line.split('')
  when 1
    # skip
  else
    match_data = MAP_REGEX.match(line)
    network[match_data[:node]] = [match_data[:left], match_data[:right]]
  end
  index += 1
end

steps = 0
current_node = 'AAA'
while current_node != 'ZZZ' do
  instruction = instructions.first
  current_node = network[current_node][instruction == 'L' ? 0 : 1]
  instructions.rotate!
  steps += 1
end

pp steps

File: 2025/day_1/part_1.py
import re
import sys

file = sys.argv[1]
break_at = int(sys.argv[2]) if len(sys.argv) > 2 else -1

counter = 0
regex = re.compile(r'(?P<direction>[LR])(?P<steps>\d+)')
current_position = 50
index = 0

with open(file + '.txt', 'r') as file:
    for line in file:
        match = regex.match(line)
        if match:
            direction = match.group('direction')
            steps = int(match.group('steps'))

            if match.group('direction') == 'L':
                current_position -= steps
            else:
                current_position += steps

            current_position = current_position % 100

            if current_position > 99:
                current_position -= 100
            elif current_position < 0:
                current_position = 100 + current_position

            print(direction, steps, current_position)

            if current_position == 0:
                counter += 1
        index += 1
        if index == break_at:
           break
print(counter)
